internal/generate/packagemanifest.go

```diff
@@ -0,0 +1,180 @@
+// Copyright 2020 The Operator-SDK Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package generate
+
+import (
+	"errors"
+	"fmt"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	apimanifests "github.com/operator-framework/api/pkg/manifests"
+	"github.com/operator-framework/api/pkg/validation"
+	log "github.com/sirupsen/logrus"
+
+	genutil "github.com/operator-framework/operator-sdk/internal/generate/internal"
+	"github.com/operator-framework/operator-sdk/internal/generate/packagemanifest/bases"
+)
+
+const (
+	// File extension for all PackageManifests written by Generator.
+	packageManifestFileExt = ".package.yaml"
+)
+
+var (
+	// User-facing errors.
+
+	// ErrNoVersion if no version # has been provided
+	ErrNoVersion = errors.New("version must be set")
+
+	// Internal errors.
+
+	// ErrNoOpName if the operator name has not been set
+	ErrNoOpName = genutil.InternalError("operator name must be set")
+	// ErrNoOutputDir if the directory to write the package manifest to has not been set
+	ErrNoOutputDir = genutil.InternalError("output directory must be set")
+)
+
+type PkgOptions struct {
+	// BaseDir is a directory to look for an existing base package manifest
+	// to update.
+	BaseDir string
+	// ChannelName is operator's PackageManifest channel. If a new PackageManifest is generated
+	// or ChannelName is the only channel in the generated PackageManifest,
+	// this channel will be set to the PackageManifest's default.
+	ChannelName string
+	// IsDefaultChannel determines whether ChannelName should be the default channel in the
+	// generated PackageManifest. If true, ChannelName will be the PackageManifest's default channel.
+	// Setting this field is only necessary when more than one channel exists.
+	IsDefaultChannel bool
+}
+
+// Generate configures the Generator with opts then runs it.
+func (g Generator) GeneratePackageManifest(operatorName, version, outputDir string, opts ...*PkgOptions) error {
+	if operatorName == "" {
+		return ErrNoOpName
+	} else if version == "" {
+		return ErrNoVersion
+	} else if outputDir == "" {
+		return ErrNoOutputDir
+	}
+	myOpts := &PkgOptions{}
+	if len(opts) > 0 {
+		myOpts = opts[0]
+	}
+
+	pkg, err := g.generatePackageManifest(operatorName, version, myOpts)
```

Comment:
IMO it shows strange. Why we have an exported and n internal func with the same name?

Reply:
This is a very common pattern.

Reply:
that bit is copied from the original interface. GeneratePackageManifest is (generate a package manifest and all the other needed junk like write it to disk. generatePackageManifest is (actually generate me a package manifest object)

Reply:
The original is not good and because of this, we are trying to refractory it. Am I right? So, let's try to shape it as better as possible. 

Regards the scenario, IMO it reduces the maintainability since we have an unambiguous representation which shows hurting principles such as DRY (Don’t Repeat Yourself).

However, I am human been and I can be wrong. Also, I love to learn something new. So, why it would be a good practice? When we should name 2 funcs with the same name? How it can help regards the readability? Could you please provide a reference that supports it as good practice? Have we any reason/argumentation that supports to adopt this practice at all?

Reply:
generatePackageManifest is a helper method to help separate the different logical parts of GeneratePackageManifest to make it easier to reason about. It's a private method so it can't be used outside of this package. I don't think the naming is too confusing but am also not particularly worried about the change. My primary intent about this refactor was to make the relationship between the cmd layer and the generator more easily testable, and removing confusing indirection like the callback functions used in the previous implementation.

Reply:
IMO this overload would be OK and is a common pattern when we want to provide an API less complexity to facilitate its usage and keep the complexity internally. However,  see: 

`GeneratePackageManifest(operatorName, version, outputDir string, opts ...*PkgOptions)` //  The Export API is more complexity 
`g.generatePackageManifest(operatorName, version, myOpts)`

And the actions that are done between the `GeneratePackageManifest` to `generatePackageManifest` are not all related to generation at all. See that it is doing validations as well which is hurting at least a few concepts such as cushion and single responsibilities.

Reply:
So what do you propose? Move the bits that do validation and write the file to disk into separate helper methods? They would still only be called internally here so it would just be moving the code around. The validation that's happening is just checking that I have the correct info to perform the method call, so it's not going to get moved out of this layer. The writing file to disk could be extracted, but that seems to be needlessly adding complexity unless we need to write packagemanifests to disk in multiple places, which I don't think we do.